### PR TITLE
modelcache: Add read and write endpoints to Redis (PROJQUAY-1939)

### PR DESCRIPTION
**Pull request by syed**
Adds `read_client` and `write_client` for Redis ModelCache to seperate out read
and write calls to separate on endpoints

### CODE REVIEW

File: data/cache/impl.py [hunk 1]
```diff
@@ -244,19 +244,32 @@ class RedisDataModelCache(DataModelCache):
...
+            self,
+            cache_config,
+            read_host="127.0.0.1",
+            read_port=6379,
+            write_host="127.0.0.1",
+            write_port=6379,
+            password=None,
+            db=0,
+            ca_cert=None,
+            ssl=False,
```
Comment (kleesc) [inline data/cache/impl.py:256]: The replica may have different config value for these parameters?

File: data/cache/impl.py [hunk 2]
```diff
@@ -244,19 +244,32 @@ class RedisDataModelCache(DataModelCache):
...
-        host="127.0.0.1",
-        port=6379,
-        password=None,
-        db=0,
-        ca_cert=None,
-        ssl=False,
+            self,
+            cache_config,
+            read_host="127.0.0.1",
+            read_port=6379,
```
Comment (kleesc) [inline data/cache/impl.py:250]: These new values should also be added to the default config.

**Review by kleesc — CHANGES_REQUESTED** [review]

**Review by kleesc — CHANGES_REQUESTED** [review]

File: data/cache/impl.py (hunk 1 above)
Comment (syed) [reply data/cache/impl.py:256]: I don't think the replica can have a different password or DB. Not sure about CA cert but I'd don't see a use case where we will have a different CA for primary and replica

File: data/cache/impl.py (hunk 2 above)
Comment (syed) [reply data/cache/impl.py:250]: The `config.py` has the Memcache settings. Can we add this without causing any issues?

File: data/cache/__init__.py [hunk 3]
```diff
@@ -38,14 +38,17 @@ def get_model_cache(config):
         return cache
 
     if engine == "redis":
-        host = cache_config.get("host", None)
-        if host is None:
-            raise Exception("Missing `host` for Redis model cache configuration")
+        read_host = cache_config.get("read_host", None)
+        write_host = cache_config.get("write_host", None)
+        if read_host is None or write_host is None:
```
[the remark this answers is not in the document]
Comment (syed) [reply data/cache/__init__.py]: 👍

File: data/cache/impl.py (hunk 2 above)
Comment (syed) [reply data/cache/impl.py]: i've added these settings in the enhancement repo https://github.com/quay/enhancements/pull/8

File: data/cache/impl.py (hunk 1 above)
Comment (kleesc) [reply data/cache/impl.py]: Not sure about passwords, but I'm assuming if a replica can be setup on a separate/completely different host, it should be able to have a separate cert for validation.
We should create a wrapper around the separate configs instead (maybe a namedtuple):
```
_RedisConfig = namedtuple("_RedisConfig", ["host", "port", "password", ...])
```

File: data/cache/impl.py (hunk 2 above)
Comment (kleesc) [reply data/cache/impl.py]: If you add/change config values, defaults should be set/updated in `config.py`.

Comment (kleesc) [reply data/cache/impl.py]: See https://github.com/quay/quay/blob/master/config.py

Comment (kleesc) [reply data/cache/impl.py]: For example, with some of the host key name changes, the values in `config.py` are no longer valid.

File: data/cache/impl.py (hunk 1 above)
Comment (syed) [reply data/cache/impl.py]: `redis-py` supports single host config only. We would never pass more than 2 hosts to the Redis ModelCache. It's simple if we just pass them via the constructor

File: data/cache/impl.py (hunk 2 above)
Comment (kleesc) [reply data/cache/impl.py]: Nevermind, the default uses memcached

File: data/cache/__init__.py [hunk 4]
```diff
@@ -38,18 +38,16 @@ def get_model_cache(config):
         return cache
 
     if engine == "redis":
-        host = cache_config.get("host", None)
-        if host is None:
-            raise Exception("Missing `host` for Redis model cache configuration")
+        primary_config = cache_config.get("primary_config", None)
+        replica_config = cache_config.get("replica_config", primary_config)
```
Comment (kleesc) [inline data/cache/__init__.py:42]: nit: remove the `_config` suffix from the config key.
The final config should look like:
```
{
  "engine":  "redis",
  "primary": {
    ...
  }
  "replica": {
    ...
  }
}
```

**Review by kleesc — CHANGES_REQUESTED** [review]

File: data/cache/impl.py [hunk 5]
```diff
@@ -243,24 +243,41 @@ class RedisDataModelCache(DataModelCache):
...
+            password=replica_config.get('password'),
+            db=replica_config.get('db'),
+            ssl_ca_certs=replica_config.get('ca_certs'),
+            ssl=replica_config.get('ssl'),
+            socket_connect_timeout=1,
+            socket_timeout=2,
+            health_check_interval=2,
+        )
+
+        self.write_client = StrictRedis(
```
[the remark this answers is not in the document]
Comment (syed) [reply data/cache/impl.py:253]: 👍

File: data/cache/__init__.py (hunk 4 above)
Comment (syed) [reply data/cache/__init__.py]: 👍

File: data/cache/impl.py [hunk 6]
```diff
@@ -243,24 +243,41 @@ class RedisDataModelCache(DataModelCache):
...
     ):
         super(RedisDataModelCache, self).__init__(cache_config)
-        self.client = StrictRedis(
-            host=host,
-            port=port,
-            password=password,
-            db=db,
-            ssl_ca_certs=ca_cert,
-            ssl=ssl,
+        self.read_client = StrictRedis(
```
[the remark this answers is not in the document]
Comment (syed) [reply data/cache/impl.py]: 👍

File: data/cache/impl.py [hunk 7]
```diff
@@ -243,24 +243,41 @@ class RedisDataModelCache(DataModelCache):
     Implementation of the data model cache backed by a Redis service.
     """
 
+    _DEFAULT_CONFIG = {
```
[the remark this answers is not in the document]
Comment (syed) [reply data/cache/impl.py]: 👍

**Review by kleesc — APPROVED** [review]
LGTM

**Pull request merged**